includes/admin/class-llms-sendwp.php

```diff
@@ -0,0 +1,399 @@
+
+		add_filter( 'lifterlms_engagements_settings', array( $this, 'add_settings' ) );
+		add_action( 'admin_print_styles', array( $this, 'output_css' ) );
+		add_action( 'admin_print_footer_scripts', array( $this, 'output_js' ) );
+		add_action( 'wp_ajax_llms_sendwp_remote_install', array( $this, 'ajax_callback_remote_install' ) );
+
+	}
+
+	/**
+	 * Add Settings.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $settings Existing settings.
+	 * @return array
+	 */
+	public function add_settings( $settings ) {
+
+		// Short circuit if missing unauthorized.
+		if ( ! current_user_can( 'install_plugins' ) ) {
+			return $settings;
+		}
+
+		$new_settings = array(
+			array(
+				'id'   => 'sendwp_connect_start',
+				'type' => 'sectionstart',
+			),
+			array(
+				'id'    => 'sendwp_title',
+				'title' => __( 'SendWP Email', 'lifterlms' ),
+				'type'  => 'title',
+				'desc'  => sprintf(
+					// Translators: %1$s = Opening anchor tag; %2$s = Closing anchor tag.
+					__( '%1$sSendWP%2$s makes WordPress email delivery as simple as a few clicks so you can relax, knowing your important emails are being delivered on time.', 'lifterlms' ),
+					'<a href="https://lifterlikes.com/sendwp" target="_blank" rel="noopener noreferrer">',
+					'</a>'
+				),
+			),
+			array(
+				'id'    => 'sendwp_connect',
+				'type'  => 'custom-html',
+				'value' => $this->get_connect_setting(),
+			),
+			array(
+				'id'   => 'sendwp_connect_end',
+				'type' => 'sectionend',
+			),
+		);
+
+		array_splice( $settings, $this->get_splice_index( $settings ), 0, $new_settings );
+
+		return $settings;
+
+	}
+
+	/**
+	 * Ajax callback for installing SendWP Plugin.
+	 *
+	 * @since [version]
+	 *
+	 * @hook wp_ajax_llms_sendwp_remote_install
+	 *
+	 * @return void
+	 */
+	public function ajax_callback_remote_install() {
+
+		$ret = $this->do_remote_install();
+		ob_clean();
+		wp_send_json( $ret, ! empty( $ret['status'] ) ? $ret['status'] : 200 );
+
+	}
+
+	/**
+	 * Remote installation method.
+	 *
+	 * @since [version]
+	 *
+	 * @return array
+	 */
+	public function do_remote_install() {
+
+		if ( ! current_user_can( 'install_plugins' ) ) {
+			return array(
+				'code'    => 'llms_sendwp_install_unauthorized',
+				'message' => __( 'You do not have permission to perform this action.', 'lifterlms' ),
+				'status'  => 403,
+			);
+		}
+
+		$install = $this->install();
+		if ( is_wp_error( $install ) ) {
+			return array(
+				'code'    => $install->get_error_code(),
+				'message' => $install->get_error_message(),
+				'status'  => 400,
+			);
+		}
+
+		return array(
+			'partner_id'      => 2007,
+			'register_url'    => sendwp_get_server_url() . '_/signup',
+			'client_name'     => sendwp_get_client_name(),
+			'client_secret'   => sendwp_get_client_secret(),
+			'client_redirect' => sendwp_get_client_redirect(),
+		);
+
+	}
+
+	/**
+	 * Install / Activate SendWP plugin.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|true
+	 */
+	private function install() {
+
```

Comment:
we could check on if `is_plugin_active( 'sendwp/sendwp.php' );` (requires: `include_once( ABSPATH . 'wp-admin/includes/plugin.php' );`) and skip all the rest in case the plugin is already active (but user not connected).

Reply:
Yes we could but activating an already active via `activate_plugin()` responds with `null` the same as if the plugin was successfully activated from a deactivated state. So I think the current version does more or less the same thing without having to add another path to the function.

Reply:
Ok I see what you mean, my thought was to skip the 
`foreach ( get_plugins() as $path => $details )`
with `get_plugins` that might be expensive, although its result is cached I'm not sure whether it's already called during a typical admin request. But anyways that's fine :D

Reply:
Oh I see what you mean... Well, I don't feel like rewriting and retesting... whatever...

I did see if anyone has benchmarked `in_array()` vs `foreach()` and found this: https://gist.github.com/Maks3w/3097685

`is_plugin_active()` uses `in_array()` under the hood. According to this person `in_array()` would be *slightly* faster when doing loose comparisons in the loop. So I think performance wise we're arguing over nothing (but that's really the best kind of argument, right?

Reply:
yeah we're arguing over nothing.
I was more talking about `get_plugins()` VS `is_plugin_active()` not about `in_array()` VS `foreach()`. But yes nevermind, if I find `get_plugins()` being slow I'll push a PR, but more likely I will not as this is not critical at all :D